### ISSUE TITLE
mc rm does not delete bucket if empty, but deletes bucket if it has files

### ISSUE THREAD

**Issue by kcboyle** [minio/minio#10373]
After some of the recent "bug fix" releases, some of the behavior we were expecting changed. In short, if we create a bucket with `mc`, and try to delete it, minio says the bucket does not exist. `mc ls` shows the bucket present. if we add anything to that bucket, and run the same `mc rm`, then the object is somehow found, and deletes appropriately. This was strange behavior, and unexpected. (We have changed our code to use `mc rb` instead of `mc rm`, but we felt it appropriate to file this issue anyway)

## Expected Behavior
I expect `mc rm` to remove my bucket, even if it is empty OR I expect `mc rm` to never remove my bucket at all

## Current Behavior
`mc rm` will only remove the bucket if there are files inside, and errors "it cannot find the object" otherwise

## Possible Solution
Unsure. Prior to Aug 17th (whatever version of minio was released then) we did not encounter this issue.

## Steps to Reproduce (for bugs)
```
○ →  mc mb my-bucket
Bucket created successfully `my-bucket`.

○ → mc rm my-bucket --recursive --force
mc: <ERROR> Failed to remove `my-bucket`. Object does not exist

○ → mc ls
[2020-06-24 13:44:08 MDT]    319B main.go
[2020-08-28 11:01:07 MDT]     64B my-bucket/

○ → touch my-bucket/file.txt

○ → mc rm my-bucket --recursive --force
Removing `my-bucket/file.txt`.
```

## Context
We were testing error cases in our codebase using minio.
We wanted to show the user a helpful error message if the bucket was empty (meaning that it did not have any matching products in it for us to use). In the test, we run
`mc mb my-bucket`
[run-test]
`mc rm my-bucket --force --recursive` 
We remove the bucket after every test to ensure a clean test environment for the next test, so contamination does not occur.

## Regression
 Yes. This was not an issue before Aug 17
 Not sure which commit caused the regression

## Your Environment

* Version used (`minio --version`): minio version RELEASE.2020-08-27T05-16-20Z
* Server setup and configuration: N/A
* Operating System and version (`uname -a`): OSX 10.15.6 AND ubuntu bionic

**Comment by harshavardhana** [comment]
`mc rm` won't delete buckets @kcboyle it is `mc rb`'s job

**Comment by kcboyle** [comment]
@harshavardhana Well, it used to.
and it will still delete the bucket if there is something in it

**Comment by harshavardhana** [comment]
> @harshavardhana Well, it used to.
> and it will still delete the bucket if there is something in it

yes that has been changed some time ago `mc rm` only deletes objects and `mc rb` is meant to delete buckets.

**Comment by harshavardhana** [comment]
`mc rb` has a bug I agree

**Comment by kcboyle** [comment]
`mc rb` is working great (and is what we needed). `mc rm` is the one that will actually delete the whole bucket if it finds files in the bucket, not just the files. 

so if `mc rm` is _not_ supposed to delete buckets, it is doing something it is not supposed to.

**Comment by harshavardhana** [comment]
> `mc rb` is working great (and is what we needed). `mc rm` is the one that will actually delete the whole bucket if it finds files in the bucket, not just the files.

mc rb deletes files and buckets
mc rm deletes only files @kcboyle leaves the bucket as is.

**Comment by harshavardhana** [comment]
NOTE: there is a difference between filesystem v/s object storage behavior from your report I see that you are running this to local FS?

**Comment by kcboyle** [comment]
We are running this on a local filesystem (in the mac OS and ubuntu cases). 

at least in a local filesystem use case, what you are saying is expected behavior is incorrect. I ran the local test again, and tried adding the following:
```
○ → mc rm my-bucket --recursive --force
Removing `my-bucket/file.txt`.

○ → mc rb --force --dangerous my-bucket
mc: <ERROR> Unable to validate target `my-bucket`. Requested file `my-bucket` not found
```

**Comment by harshavardhana** [comment]
Local filesystem has no concept of buckets so it doesn't know @kcboyle

**Comment by harshavardhana** [comment]
But it is possible to keep it consistent such that `mc rm` doesn't delete the parent top-level prefix.

**Comment by kcboyle** [comment]
This is behavior we noticed while solving an unexpected issue in our running CI. `mc rm` did not fail for us before if the bucket had no contents, and it does now. 
We have changed it to `mc rb`, which was actually the correct way to do it, we realize, and are unblocked. 

If you don't think this behavior is an issue, feel free to close.

**Comment by harshavardhana** [comment]
> This is behavior we noticed while solving an unexpected issue in our running CI. `mc rm` did not fail for us before if the bucket had no contents, and it does now. 
> We have changed it to `mc rb`, which was actually the correct way to do it, we realize, and are unblocked. 
> 
> If you don't think this behavior is an issue, feel free to close.

Will get back..
